src/cpp/shared/game/game.cpp

```diff
@@ -165,4 +182,77 @@ void philAI::RedistributeTroops(armyGroup *army1, armyGroup *army2, int a1, int
 	if (gpGame->allowAIArmySharing) {
 		RedistributeTroops_orig(army1, army2, a1, a2, a3, a4, a5);
 	}
+}
+
+void game::SetRandomHeroArmies(int heroIdx, int isAI) {
```

Comment:
Ignore the other comments. This entire function can be greatly simplified in structure.

It should look something like:

ClearHeroArmy();

{
  // Tier 1 creatures
  int creatureType = ..., qty = ... ;
  GrantArmy(hero.army, creatureType, qty);
}

if (randomThing) {
  // Tier 2 creatures
  int creatureType = ..., qty = ... ;
  GrantArmy(hero.army, creatureType, qty);
}


Does this make sense?

Reply:
I see how a ClearHeroArmy function would be beneficial here and I suppose we probably should just use GrantArmy, since it's supposed to do the same thing as what happens at the end. I noticed that you suggested dividing the logic into to groups based on Tier level. This would require duplicating the most of the logic of the function for both the Tier 1 and Tier 2 groups. It would really only get rid of the FOR loop that iterates from Tier 1 to Tier 2. Is this still what you would suggest??

Reply:
Sorry, I didn't mean to actually suggest that be a separate function. I mean, that's a fine thing to do, but I just meant ClearHeroArmy() as a placeholder for whatever code to do that.

My read of this code is that it's actually using fairly separate logic for tier 1 and tier 2 creatures, and does not really gain anything from the loop. So, separating them makes the code more clear, and reduces the amount of state visible at any point.

Reply:
Okay, I had a second read. So, I guess it is using a lot of common logic. 

Instead, here are my suggestions for simplifying the function:

* Use a ClearArmy function. I think this already exists.
* Use a GrantArmy function, which definitely already exists under some name.
* Create a temporary variable for creatureTypesAndQuantities[gpGame->heroes[heroIdx].factionID]
* Shorten variable names, and move variable declarations closer to their use
* Inline some of the random numbers instead of saving to a variable

I also recommend reading this article: https://hackernoon.com/trust-no-one-an-introduction-to-large-codebases-for-new-engineers-100586a87a5f . Then you'll see why I think the comments are doing more harm than good.

With all these suggestions, I think you'll get this function to about half the current length.

Reply:
Finally, reflect on these suggestions. I want you to get to the point where things like this are obvious. If you can make your code half as long while still doing the same thing, you will be loved.

Reply:
"Create a temporary variable for creatureTypesAndQuantities[gpGame->heroes[heroIdx].factionID]". So, a pointer to the element at creatureTypesAndQuantities[gpGame->heroes[heroIdx].factionID][0][0]? I'm not sure I understand.

Reply:
Yes, except that, if you chance the elements of creatureTypesAndQuantities[x] to a structure instead of a 2D array, then it's just a pointer to the structure at creatureTypesAndQuantities[gpGame->heroes[heroIdx].factionID] . Do the structure refactoring first.

Reply:
I'm not entirely sure why it wasn't working last time--I'll acknowledge that there were more unnecessary complexities to the structure definition then--but the suggested solution is up and working.